### PR TITLE
Feature 209 map view location search

### LINKED ISSUE #209: Map view location search
As a user I need to be able to search locations from the map view and set them as starting point or end point.

### PR DESCRIPTION
**Pull request by Chi-Chu319**
resolves #209

### CODE REVIEW

File: src/components/layout/location-search.tsx [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+  /**
+   * Fetches search histories from secure storage 
+   */
+  const fetchSearchHistories = async () => {
+    try {
+      const fetchedHistories = await SecureStore.getItemAsync(LOCATION_SEARCH_HISTORIES);
+
+      if (fetchedHistories) {
+        const parsedHistories = JSON.parse(fetchedHistories) as AutocompleteOption[];
+        setLocationSearchHistories(parsedHistories);
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:72]:
```suggestion
        setLocationSearchHistories(JSON.parse(fetchedHistories));
```

File: src/components/layout/location-search.tsx [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+      return fetchedUser.homeAddress;
+    } catch (error) {
+      errorContext.setError(strings.errorDialog.getUSer, error);
+    }
+  };
+
+  /**
+   * Fetches user home address 
+   */
+  const fetchUserHomeAddress = async (userHomeAddress?: Address) => {
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:100]:
```suggestion
  /**
   * Fetches user home address
   *
   * @param userHomeAddress user home address
   */
  const fetchUserHomeAddress = async (userHomeAddress?: Address) => {
```

File: src/components/layout/location-search.tsx [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+
+    try {
+      const geocodingResponse = await GeocodingUtils.fromAddress(userHomeAddress.streetAddress);
+      const geocodingResult = geocodingResponse.results[0];
+      const coordinates = geocodingResult.geometry.location;
+      const sourceInputCoordinates: LatLng = {
+        latitude: coordinates.lat,
+        longitude: coordinates.lng
+      };
+      setUserHomeLocation(sourceInputCoordinates);
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:113]:
```suggestion
      setUserHomeLocation({
        latitude: coordinates.lat,
        longitude: coordinates.lng
      });
```

File: src/components/layout/location-search.tsx [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+      const geocodingResponse = await GeocodingUtils.fromAddress(locationName);
+
+      const geocodingResult = geocodingResponse.results[0];
+      const autocompleteOption: AutocompleteOption = {
+        locationName: locationName,
+        latitude: geocodingResult.geometry.location.lat,
+        longitude: geocodingResult.geometry.location.lng
+      };
+
+      await onAutoCompleteOptionPress(autocompleteOption);
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:172]:
```suggestion

      await onAutoCompleteOptionPress({
        locationName: locationName,
        latitude: geocodingResult.geometry.location.lat,
        longitude: geocodingResult.geometry.location.lng
      });
```

File: src/components/layout/location-search.tsx [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+
+  /**
+   * Handler for autocomplete option press
+   * 
+   * @param autocompleteOption autocomplete option 
+   */
+  const onAutoCompleteOptionPress = async (autocompleteOption: AutocompleteOption) => {
+    const location: LatLng = { latitude: parseFloat(autocompleteOption.latitude), longitude: parseFloat(autocompleteOption.longitude) };
+    setLocationSearchInput(autocompleteOption.locationName);
+    onLocationPress(location);
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:186]:
```suggestion
    setLocationSearchInput(autocompleteOption.locationName);
    onLocationPress({
      latitude: parseFloat(autocompleteOption.latitude),
      longitude: parseFloat(autocompleteOption.longitude)
    });
```

File: src/components/layout/location-search.tsx [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+  const onLocationPress = (location: LatLng) => {
+    updateRouteAlts([]);
+    toggleListItem(false);
+    updateCurrentView(location);
+    updateNewMarker(location);
+  };
+
+  const bodyHeight = animatedController.interpolate({
+    inputRange: [0, 1],
+    outputRange: [0, bodySectionHeight]
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:214]:
```suggestion
    inputRange: [ 0, 1 ],
    outputRange: [ 0, bodySectionHeight ]
```

File: src/components/layout/location-search.tsx [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+  });
+
+  /**
+   * Toggle list open/closed animation
+   */
+  const toggleListItem = (newOpen: boolean) => {
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:220]:
```suggestion
  /**
   * Toggle list open/closed animation
   *
   * @param newOpen new open value
   */
  const toggleListItem = (newOpen: boolean) => {
```

File: src/components/layout/location-search.tsx [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+   * @param onLocationPress on location press handler
+   */
+  const renderSearchResult = (locationName: string, onLocationResultPress: () => void, key: number) => (
+    <List.Item
+      key={ key }
+      style={ styles.locationItem }
+      titleStyle={ styles.locationItemTitle }
+      title={ locationName }
+      onPress={ onLocationResultPress }
+      left={ props => <List.Icon {...props} color="#417a8f" style={ styles.locationItemIcon } icon="map-marker"/> }
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:364]:
```suggestion
      left={ props =>
        <List.Icon
          { ...props }
          color="#417a8f"
          style={ styles.locationItemIcon }
          icon="map-marker"
        />
      }
```

File: src/components/layout/location-search.tsx [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+      renderSearchResult(locationSearchResult, () => onLocationSearchResultPress(locationSearchResult), index));
+  };
+
+  /**
+   * Renders location search result
+   * 
+   * @param locationName location name
+   * @param onLocationPress on location press handler
+   */
+  const renderSearchResult = (locationName: string, onLocationResultPress: () => void, key: number) => (
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:357]:
```suggestion
  /**
   * Renders location search result
   * 
   * @param locationName location name
   * @param onLocationPress on location press handler
   * @param key unique key for rendered item
   */
  const renderSearchResult = (locationName: string, onLocationResultPress: () => void, key: number) => (
```

File: src/components/layout/location-search.tsx [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+          <ActivityIndicator
+            color="#8fcfd1"
+            size={ 40 }
+          />
+        </View>
+      );
+    }
+
+    return locationSearchResults.map((locationSearchResult, index) =>
+      renderSearchResult(locationSearchResult, () => onLocationSearchResultPress(locationSearchResult), index));
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:348]:
```suggestion
    return locationSearchResults.map((locationSearchResult, index) =>
      renderSearchResult(
        locationSearchResult,
        () => onLocationSearchResultPress(locationSearchResult),
        index
      )
    );
```

File: src/components/layout/location-search.tsx [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+  );
+
+  /**
+   * Render recent searches
+   */
+  const renderSearchHistories = () => (
+    <View>
+      <Text style={ styles.recentSearchesTitle }>{ strings.mapScreen.recentSearches }</Text>
+      { locationSearchHistories.map((locationSearchHistory, index) =>
+        renderSearchResult(locationSearchHistory.locationName, () => onAutoCompleteOptionPress(locationSearchHistory), index)) }
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:375]:
```suggestion
      <Text style={ styles.recentSearchesTitle }>
        { strings.mapScreen.recentSearches }
      </Text>
      { locationSearchHistories.map((locationSearchHistory, index) =>
        renderSearchResult(
          locationSearchHistory.locationName,
          () => onAutoCompleteOptionPress(locationSearchHistory),
          index
        )
      )}
```

File: src/components/layout/location-search.tsx [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+            top: 10,
+            height: bodyHeight
+          }}
+        >
+          <ScrollView
+            scrollEventThrottle={ 32 }
+            onScroll={ () => Keyboard.dismiss() }
+            onLayout={ event =>
+              setBodySectionHeight(event.nativeEvent.layout.height)
+            }
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:432]:
```suggestion
            onLayout={ event => setBodySectionHeight(event.nativeEvent.layout.height) }
```

File: src/components/layout/location-search.tsx [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+                flex: 1,
+                justifyContent: "center",
+                alignItems: "center"
+              }}
+            >
+              <LinearGradient
+                style={{ height: "100%", width: "100%" }}
+                colors={ ["#fff", "transparent"]}
+                start={ [0.95, 0] }
+                end={ [1, 0] }
```
Comment (lauriLukkarinen) [inline src/components/layout/location-search.tsx:459]:
```suggestion
                colors={[ "#fff", "transparent" ]}
                start={[ 0.95, 0 ]}
                end={[ 1, 0 ]}
```

File: src/components/map.tsx [hunk 14]
```diff
@@ -29,43 +26,38 @@ interface Props {
...
+  newFavouriteLocationName: string;
 }
 
 /**
  * Map component
  */
 class Map extends React.Component<Props, State> {
-
-  mapRef: React.RefObject<MapView>;
+
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:59]:
```suggestion

```

File: src/components/map.tsx [hunk 15]
```diff
@@ -203,106 +161,25 @@ class Map extends React.Component<Props, State> {
...
-              { strings.custom }
-            </Button>
-          </View>
-        </>
-      );
-    }
-    return null;
+  private renderNewFavouriteLocationDialog = () => {
+    const { savingLocation, updateSavingLocation } = this.props;
+    const { newFavouriteLocationName } = this.state;
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:168]:
```suggestion
    const { newFavouriteLocationName } = this.state;

```

File: src/components/map.tsx [hunk 16]
```diff
@@ -346,124 +223,103 @@ class Map extends React.Component<Props, State> {
...
-          >
-            <MapMarkerNew fill="#277B90"/>
-          </Marker>
-        </>
+        <Marker
+          coordinate={{
+            latitude: newMarker.latitude,
+            longitude: newMarker.longitude
+          }}
+          draggable={ true }
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:239]:
```suggestion
          draggable
```

File: src/components/map.tsx [hunk 17]
```diff
@@ -346,124 +223,103 @@ class Map extends React.Component<Props, State> {
...
-      efficientDisabled: true,
-      relaxedDisabled: true,
-      customDisabled: true
+      newFavouriteLocationName: newFavouriteLocationName
     });
   };
 
   /**
-   * Set efficient value
+   * On Set destination button press
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:258]:
```suggestion
   * On save location confirm
```

File: src/components/map.tsx [hunk 18]
```diff
@@ -346,124 +223,103 @@ class Map extends React.Component<Props, State> {
...
+    const { setError } = this.context;
 
-  /**
-   * Set strict value
-   */
-  private setRelaxed = () => {
+    if (!newMarker || !accessToken) {
+      return;
+    }
+    let favouriteLocationAddress: Address | undefined;
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:275]:
```suggestion

    let favouriteLocationAddress: Address | undefined;
```

File: src/components/map.tsx [hunk 19]
```diff
@@ -346,124 +223,103 @@ class Map extends React.Component<Props, State> {
...
+      const geocodingResponse = await GeocodingUtils.fromLatLng(newMarker.latitude, newMarker.longitude);
+
+      const geocodingResultComponents = geocodingResponse.results[0]?.address_components;
+      favouriteLocationAddress = geocodingResultComponents && {
+        streetAddress:
+          `${geocodingResultComponents.find((component: any) => component.types.includes("route"))?.short_name} ${geocodingResultComponents.find((component: any) => component.types.includes("street_number"))?.short_name}`,
+        city: geocodingResultComponents.find((component: any) => component.types.includes("locality"))?.short_name,
+        postalCode: geocodingResultComponents.find((component: any) => component.types.includes("postal_code"))?.short_name,
+        country: geocodingResultComponents.find((component: any) => component.types.includes("country"))?.long_name
+      };
```
Comment (lauriLukkarinen) [inline src/components/map.tsx:287]:
```suggestion
      const route = geocodingResultComponents.find((component: any) => component.types.includes("route"));
      const streetNumber = geocodingResultComponents.find((component: any) => component.types.includes("street_number"));
      const city = geocodingResultComponents.find((component: any) => component.types.includes("locality"));
      const postalCode = geocodingResultComponents.find((component: any) => component.types.includes("postal_code"));
      const country = geocodingResultComponents.find((component: any) => component.types.includes("country"));

      favouriteLocationAddress = !!geocodingResultComponents && {
        streetAddress: `${route?.short_name || ""} ${streetNumber?.short_name || ""}`,
        city: city?.short_name,
        postalCode: postalCode?.short_name,
        country: country?.long_name
      };
```

File: src/components/screens/main-screen.tsx [hunk 20]
```diff
@@ -121,8 +120,8 @@ class MainScreen extends React.Component<Props, State> {
 
     if (accessToken && prevProps.accessToken !== accessToken) {
       this.getFavouriteLocations();
-    } else if (!accessToken && favouriteLocationData.length > 0) {
-      this.setFavouriteLocationDataEmpty();
+    } else if (!accessToken && favouriteLocations.length > 0) {
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:123]:
```suggestion
    } else if (!accessToken && favouriteLocations.length) {
```

File: src/components/screens/main-screen.tsx [hunk 21]
```diff
@@ -177,48 +174,61 @@ class MainScreen extends React.Component<Props, State> {
...
+                updateSavingLocation={ this.updateSavingLocation }
+              />
+              { Platform.OS === "ios" ?
+                <KeyboardAvoidingView
+                  behavior="position"
+                >
+                  { this.renderFooter() }
+                </KeyboardAvoidingView>
+                :
+                this.renderFooter()
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:218]:
```suggestion
              { Platform.OS === "ios" ?
                <KeyboardAvoidingView behavior="position">{ this.renderFooter() }</KeyboardAvoidingView> :
                this.renderFooter()
```

File: src/components/screens/main-screen.tsx [hunk 22]
```diff
@@ -467,6 +523,15 @@ class MainScreen extends React.Component<Props, State> {
     return names;
   };
 
+  /**
+   * Update show footer content
+   */
+  private setShowFooterContents = (newShowFooterContent: boolean) => {
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:529]:
```suggestion
  /**
   * Update show footer content
   *
   * @param newShowFooterContent new show footer content value
   */
  private setShowFooterContents = (newShowFooterContent: boolean) => {
```

File: src/components/screens/main-screen.tsx [hunk 23]
```diff
@@ -467,6 +523,15 @@ class MainScreen extends React.Component<Props, State> {
...
+    this.setState({
+      showFooterContents: newShowFooterContent
+    });
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:532]:
```suggestion
    this.setState({ showFooterContents: newShowFooterContent });
```

File: src/components/screens/main-screen.tsx [hunk 24]
```diff
@@ -482,8 +547,15 @@ class MainScreen extends React.Component<Props, State> {
...
-    this.setState({ currentView: currentView });
+  private updateCurrentView = async (currentView?: LatLng) => {
+    if (!this.mapRef.current || !currentView) {
+      return;
+    }
+    await this.mapRef.current.animateCamera({
+      center: currentView,
+      zoom: 12,
+      altitude: 50000,
+    }, { duration: 2000 });
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:558]:
```suggestion
  private updateCurrentView = (currentView?: LatLng) => {
    if (this.mapRef.current && currentView) {
      await this.mapRef.current.animateCamera({
        center: currentView,
        zoom: 12,
        altitude: 50000,
      }, { duration: 2000 });
    }
```

File: src/components/screens/main-screen.tsx [hunk 25]
```diff
@@ -512,6 +584,17 @@ class MainScreen extends React.Component<Props, State> {
...
 
+  /**
+   * Updates favourite location
+   * 
+   * @param favouriteLocations favourite location array
+   */
+  private updateFavouriteLocations = (favouriteLocations: FavouriteLocation[]) => {
+    this.setState({
+      favouriteLocations: favouriteLocations
+    });
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:595]:
```suggestion
    this.setState({ favouriteLocations: favouriteLocations });
```

File: src/components/screens/main-screen.tsx [hunk 26]
```diff
@@ -846,31 +925,36 @@ class MainScreen extends React.Component<Props, State> {
...
-  }
+  };
 
   /**
    * Update selected routing mode
    */
+  private updateRouteAlts = (routeAlts: RouteAlternative[]) => {
+    this.setState({
+      routeAlts: routeAlts
+    });
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:946]:
```suggestion
    this.setState({ routeAlts: routeAlts });
```

File: src/components/screens/main-screen.tsx [hunk 27]
```diff
@@ -846,31 +925,36 @@ class MainScreen extends React.Component<Props, State> {
...
-      loadingRouteAirQuality: false
+      routeAlts: updatedRouteAlts
     });
-  }
+  };
 
   /**
    * Update selected routing mode
    */
+  private updateRouteAlts = (routeAlts: RouteAlternative[]) => {
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:943]: Fix tsdocs

File: src/components/screens/main-screen.tsx [hunk 28]
```diff
@@ -893,8 +977,64 @@ class MainScreen extends React.Component<Props, State> {
...
+
+    if (!newMarker) {
+      return;
+    }
+
+    const newDestination = { ...newMarker };
+
+    try {
+      const locationGeocodedName = await this.getLocationAddress(newMarker.latitude.toString(), newMarker.longitude.toString());
+      this.updateDestinationCoordinates(newDestination, locationGeocodedName);
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:1000]:
```suggestion
    try {
      const locationGeocodedName = await this.getLocationAddress(newMarker.latitude.toString(), newMarker.longitude.toString());
      this.updateDestinationCoordinates({ ...newMarker }, locationGeocodedName);
```

File: src/components/screens/main-screen.tsx [hunk 29]
```diff
@@ -893,8 +977,64 @@ class MainScreen extends React.Component<Props, State> {
...
+
+    if (!newMarker) {
+      return;
+    }
+
+    const newOrigin = { ...newMarker };
+
+    try {
+      const locationGeocodedName = await this.getLocationAddress(newMarker.latitude.toString(), newMarker.longitude.toString());
+      this.updateSourceCoordinates(newOrigin, locationGeocodedName);
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:1023]:
```suggestion
    try {
      const locationGeocodedName = await this.getLocationAddress(newMarker.latitude.toString(), newMarker.longitude.toString());
      this.updateSourceCoordinates({ ...newMarker }, locationGeocodedName);
```

File: src/components/screens/main-screen.tsx [hunk 30]
```diff
@@ -893,8 +977,64 @@ class MainScreen extends React.Component<Props, State> {
...
+    this.updateNewMarker(undefined);
+  };
+
+  /**
+   * Update saving location state
+   */
+  private updateSavingLocation = (updatedSavingLocation: boolean) => {
+    this.setState({
+      savingLocation: updatedSavingLocation
+    });
```
Comment (lauriLukkarinen) [inline src/components/screens/main-screen.tsx:1037]:
```suggestion
  /**
   * Update saving location state
   *
   * @param updatedSavingLocation updated saving location state
   */
  private updateSavingLocation = (updatedSavingLocation: boolean) => {
    this.setState({ savingLocation: updatedSavingLocation });
```

File: src/components/screens/navigation-screen.tsx [hunk 31]
```diff
@@ -312,17 +311,16 @@ class NavigationScreen extends React.Component<Props, State> {
...
           <View style={ styles.stepIconContainer }>
             { this.relativeDirectionInterpreter(secondStep.relativeDirection) }
           </View>
         }
         style={ styles.nextStreet }
         titleStyle={ styles.nextStreetName }
         descriptionStyle={ styles.streetDistance }
-        onPress={ () => { this.setState({ expanded: !expanded }), LayoutAnimation.easeInEaseOut() }}
-        right={ props => 
+        onPress={ () => { this.setState({ expanded: !expanded }); LayoutAnimation.easeInEaseOut(); }}
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:322]:
```suggestion
        onPress={ () => {
          this.setState({ expanded: !expanded });
          LayoutAnimation.easeInEaseOut();
        }}
```

File: src/components/screens/navigation-screen.tsx [hunk 32]
```diff
@@ -338,41 +336,37 @@ class NavigationScreen extends React.Component<Props, State> {
         </ScrollView>
       </List.Accordion>
     );
-  }
+  };
 
   /**
    * Render navigation step item
    */
   private renderRemainingStepsList = (remainingSteps: NavigationStep[]) => {
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:344]:
```suggestion
  /**
   * Render navigation step item
   *
   * @param remainingSteps remaining navigation steps
   */
  private renderRemainingStepsList = (remainingSteps: NavigationStep[]) => {
```

File: src/components/screens/navigation-screen.tsx [hunk 33]
```diff
@@ -413,20 +407,19 @@ class NavigationScreen extends React.Component<Props, State> {
...
+
+      return (
+        <IconButton
+          size={ 36 }
+          icon="crosshairs-gps"
+          onPress={ () =>
+            this.setState({
+              followsUserLocation: false
+            })
+          }
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:420]:
```suggestion
          onPress={ () => this.setState({ followsUserLocation: false }) }
```

File: src/components/screens/navigation-screen.tsx [hunk 34]
```diff
@@ -481,9 +474,9 @@ class NavigationScreen extends React.Component<Props, State> {
...
     
     const { latitude, longitude } = currentUserLocation.coords;
-    let matchingDeltas: PointCoordinateDelta[] = [];
+    const matchingDeltas: PointCoordinateDelta[] = [];
 
-    coordinateDeltaPoints.map(delta => {
+    coordinateDeltaPoints.forEach(delta => {
       if (delta.minLat < latitude && delta.maxLat > latitude && delta.minLon < longitude && delta.maxLon > longitude) {
         matchingDeltas.push(delta);
       }
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:482]:
```suggestion
      if (
        delta.minLat < latitude &&
        delta.maxLat > latitude &&
        delta.minLon < longitude &&
        delta.maxLon > longitude
      ) {
        matchingDeltas.push(delta);
      }
```

File: src/components/screens/navigation-screen.tsx [hunk 35]
```diff
@@ -563,7 +556,7 @@ class NavigationScreen extends React.Component<Props, State> {
     this.setState({
       coordinateDeltaPoints: coordinateDeltaPoints
     });
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:558]:
```suggestion
    this.setState({ coordinateDeltaPoints: coordinateDeltaPoints });
```

File: src/components/screens/navigation-screen.tsx [hunk 36]
```diff
@@ -587,7 +580,7 @@ class NavigationScreen extends React.Component<Props, State> {
     this.setState({
       magnetometerPermissionsRecieved: true
     });
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:582]:
```suggestion
    this.setState({ magnetometerPermissionsRecieved: true });
```

File: src/components/screens/navigation-screen.tsx [hunk 37]
```diff
@@ -628,13 +621,13 @@ class NavigationScreen extends React.Component<Props, State> {
     const angleRounded = Math.round(angle);
     const degree = angleRounded - 90 >= 0 ? angleRounded - 90 : angleRounded + 271;
     
-    //TODO: check if there are other ways to make this smooth
-    if (trackingEnabled && degree % 2 == 0) {
+    // TODO: check if there are other ways to make this smooth
+    if (trackingEnabled && degree % 2 === 0) {
       this.mapRef.current?.setCamera({
         heading: degree
       });
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:628]:
```suggestion
      this.mapRef.current?.setCamera({ heading: degree });
```

File: src/components/screens/navigation-screen.tsx [hunk 38]
```diff
@@ -678,11 +671,11 @@ class NavigationScreen extends React.Component<Props, State> {
 
     if (currentNavigationStep) {
       this.mapRef.current?.animateCamera({
-        center: {latitude: currentNavigationStep.lat, longitude: currentNavigationStep.lon},
+        center: { latitude: currentNavigationStep.lat, longitude: currentNavigationStep.lon },
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:674]:
```suggestion
        center: {
          latitude: currentNavigationStep.lat,
          longitude: currentNavigationStep.lon
        },
```

File: src/components/screens/navigation-screen.tsx [hunk 39]
```diff
@@ -699,7 +692,7 @@ class NavigationScreen extends React.Component<Props, State> {
     this.setState({
       stepRemainingDistance: remainingDistance
     });
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:694]:
```suggestion
    this.setState({ stepRemainingDistance: remainingDistance });
```

File: src/components/screens/navigation-screen.tsx [hunk 40]
```diff
@@ -766,14 +759,14 @@ class NavigationScreen extends React.Component<Props, State> {
     this.setState({
       currentNavigationStep: steps[0]
     });
```
Comment (lauriLukkarinen) [inline src/components/screens/navigation-screen.tsx:761]:
```suggestion
    this.setState({ currentNavigationStep: steps[0] });
```

File: src/utils/geocoding-utils.tsx [hunk 41]
```diff
@@ -0,0 +1,46 @@
...
+   * Geocoding api from place
+   *
+   * @param address given request parameters
+   * @returns json result
+   */
+  public static fromAddress = async (address: string): Promise<any> => {
+    try {
+      const url = `${this.GEOCODE_API}?address=${encodeURIComponent(address)}${this.parsedParams}`;
+
+      const response = await fetch(url);
```
Comment (lauriLukkarinen) [inline src/utils/geocoding-utils.tsx:21]:
```suggestion
      const response = await fetch(`${this.GEOCODE_API}?address=${encodeURIComponent(address)}${this.parsedParams}`);
```

File: src/utils/geocoding-utils.tsx [hunk 42]
```diff
@@ -0,0 +1,46 @@
...
+   *
+   * @param address given request parameters
+   * @returns json result
+   */
+  public static fromLatLng = async (lat: number | string, lng: number | string): Promise<any> => {
+    try {
+      const latLng = `${lat},${lng}`;
+      const url = `${this.GEOCODE_API}?address=${encodeURIComponent(latLng)}${this.parsedParams}`;
+
+      const response = await fetch(url);
```
Comment (lauriLukkarinen) [inline src/utils/geocoding-utils.tsx:39]:
```suggestion
      const address = encodeURIComponent(`${lat},${lng}`);
      const response = await fetch(`${this.GEOCODE_API}?address=${address}${this.parsedParams}`);
```

File: src/utils/place-utils.tsx [hunk 43]
```diff
@@ -0,0 +1,48 @@
...
+   * Place api autocomplete
+   *
+   * @param input given request parameters
+   * @returns json result
+   */
+  public static autocomplete = async (input: string): Promise<any> => {
+    try {
+      const url = `${this.PLACE_AUTOCOMPLETE_API}?input=${encodeURIComponent(input)}${this.apiKeyParam}${this.parsedParams}`;
+
+      const response = await fetch(url);
```
Comment (lauriLukkarinen) [inline src/utils/place-utils.tsx:24]:
```suggestion
      const response = await fetch(`${this.PLACE_AUTOCOMPLETE_API}?input=${encodeURIComponent(input)}${this.apiKeyParam}${this.parsedParams}`);
```

File: src/utils/place-utils.tsx [hunk 44]
```diff
@@ -0,0 +1,48 @@
...
+   * Place api place detail
+   *
+   * @param placeId given request parameters
+   * @returns json result
+   */
+  public static placeDetail = async (placeId: string): Promise<any> => {
+    try {
+      const url = `${this.PLACE_DETAILS_API}?place_id=${encodeURIComponent(placeId)}${this.apiKeyParam}`;
+
+      const response = await fetch(url);
```
Comment (lauriLukkarinen) [inline src/utils/place-utils.tsx:41]:
```suggestion
      const response = await fetch(`${this.PLACE_DETAILS_API}?place_id=${encodeURIComponent(placeId)}${this.apiKeyParam}`);
```

File: src/utils/routing-utils.tsx [hunk 45]
```diff
@@ -72,32 +71,32 @@ export default class RoutingUtils {
...
+      const pollutantFound = pollutants.find(pollutant => pollutant.id === pollutantPenalty.pollutantId);
+      if (!pollutantFound) {
+        return;
       }
-    )
 
-    return Object.keys(pollutantQueriesRaw).map(pollutantQueryName => RoutingUtils.constructQuery(pollutantQueryName, pollutantQueriesRaw[pollutantQueryName].toString())) 
-  }
+      (pollutantQueriesRaw[pollutantFound.requestParameters.thresholdRequestParameter] = pollutantPenalty.threshold);
+      (pollutantQueriesRaw[pollutantFound.requestParameters.penaltyRequestParameter] = pollutantPenalty.penalty);
```
Comment (lauriLukkarinen) [inline src/utils/routing-utils.tsx:94]:
```suggestion
      pollutantQueriesRaw[pollutantFound.requestParameters.thresholdRequestParameter] = pollutantPenalty.threshold;
      pollutantQueriesRaw[pollutantFound.requestParameters.penaltyRequestParameter] = pollutantPenalty.penalty;
```

File: src/utils/routing-utils.tsx [hunk 46]
```diff
@@ -176,11 +174,20 @@ export default class RoutingUtils {
...
+    const recommendation = selectedPollutant.recommendation || 0;
+
+    return pollutionEntryArray
+      .map(pollutionEntry =>
+        pollutionEntry?.value
+        ?
+        RoutingUtils.getPollutantColorFaction(baseColor, (pollutionEntry?.value - recommendation) / (limit - recommendation))
+        :
+        baseColor
+      );
```
Comment (lauriLukkarinen) [inline src/utils/routing-utils.tsx:190]:
```suggestion
    return pollutionEntryArray.map(pollutionEntry => {
      if (!pollutionEntry?.value) {
        return baseColor;
      }

      return RoutingUtils.getPollutantColorFaction(
        baseColor,
        (pollutionEntry.value - recommendation) / (limit - recommendation)
      );
    });
```

File: src/utils/routing-utils.tsx [hunk 47]
```diff
@@ -176,11 +174,20 @@ export default class RoutingUtils {
   public static stokeColorConverter = (pollutants: Pollutant[], routeAirQuality: AirQuality[], baseColor: string) =>{
     // TODO change the pollutant to a general air quality measurement 
     const selectedPollutant = pollutants[0];
-    const pollutionEntryArray = routeAirQuality.map(airQuality => airQuality.pollutionValues.find(pollutionValue => pollutionValue.pollutantId === selectedPollutant.id));
-
-    const limit = selectedPollutant.limit|| 0;
-    const recommendation = selectedPollutant.recommendation|| 0;
+    const pollutionEntryArray = routeAirQuality
+      .map(airQuality => airQuality.pollutionValues.find(pollutionValue => pollutionValue.pollutantId === selectedPollutant.id));
```
Comment (lauriLukkarinen) [inline src/utils/routing-utils.tsx:178]:
```suggestion
    const pollutionEntryArray = routeAirQuality.map(
      ({ pollutionValues }) => pollutionValues.find(pollutionValue => pollutionValue.pollutantId === selectedPollutant.id)
    );
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/layout/footer-route-options.tsx [hunk 48]
```diff
@@ -344,47 +369,43 @@ class FooterRouteOptions extends React.Component<Props, State> {
...
+      const geocodingResponse = await GeocodingUtils.fromAddress(address);
 
-        this.setState({
-          autocompleteOptions: autocompleteOptions
-        });
-      } catch (error) {
-        const { setError } = this.context;
-        setError(strings.errorDialog.geocodeSuggestionError, error);
-      }
+      const autocompleteOptions: AutocompleteOption[] = geocodingResponse.results.map((item: any) => ({
```
[the remark this answers is not in the document]
Comment (Chi-Chu319) [reply src/components/layout/footer-route-options.tsx:384]: We dont have type support from google map API, and I just didn't bother to make our own ones...

**Review by Chi-Chu319 — COMMENTED** [review]
comment

File: src/components/screens/main-screen.tsx (hunk 27 above)
Comment (lauriLukkarinen) [reply src/components/screens/main-screen.tsx:935]: This still needs an update

**Review by lauriLukkarinen — APPROVED** [review]

File: src/components/screens/main-screen.tsx (hunk 27 above)
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:935]:
```suggestion
  /**
   * Update selected routing mode
   *
   * @param routeAlts route alternatives
   */
  private updateRouteAlts = (routeAlts: RouteAlternative[]) => {
```

**Pull request merged**